source/rules/tokens.rs

```diff
@@ -0,0 +1,222 @@
+// Tagua VM
+//
+//
+// New BSD License
+//
+// Copyright © 2016-2016, Ivan Enderlin.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of the Hoa nor the names of its contributors may be
+//       used to endorse or promote products derived from this software without
+//       specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+//! Group of token rules.
+//!
+//! The list of all tokens is provided by the PHP Language Specification in
+//! the [Grammar chapter, Tokens
+//! section](https://github.com/php/php-langspec/blob/master/spec/19-grammar.md#tokens).
+
+use super::super::tokens;
+use super::super::ast::Name;
+
+named!(
+    pub variable<Name>,
+    map_res!(
+        preceded!(
+            tag!(tokens::VARIABLE),
+            name
+        ),
+        variable_mapper
+    )
+);
+
+fn variable_mapper(string: &[u8]) -> Result<Name, ()> {
+    Ok(Name::Variable(string))
+}
+
+named!(
+    pub qualified_name<Name>,
+    chain!(
+        head: alt!(
+            tag!(tokens::NAMESPACE_SEPARATOR)
+          | terminated!(
+                tag!(tokens::NAMESPACE),
+                tag!(tokens::NAMESPACE_SEPARATOR)
+            )
+        )? ~
+        mut accumulator: map_res!(
+            and_not!(name, tag!(tokens::NAMESPACE)),
+            wrap_into_vector_mapper
+        ) ~
+        many0!(
+            tap!(
+                tail: preceded!(
+                    tag!(tokens::NAMESPACE_SEPARATOR),
+                    and_not!(name, tag!(tokens::NAMESPACE))
+                ) =>
+                    accumulator.push(tail)
+            )
+        ),
+        || {
+            match head {
+                Some(handle) =>
```

Comment:
It would be nicer to have:
```rust
Some(tokens::NAMESPACE_SEPARATOR) => Name::FullyQualified(accumulator),
Some(_) => Name::RelativeQualified(accumulator),
None => …
```
but there is a bug in the Rust MIR.